### PR TITLE
Gas optimisation and private constant explicitly named

### PR DESCRIPTION
**Pull request by thedarkjester**
Utilises the following techniques:

1. Conversion of fixed variables to constants
2. Use default value initialisation for zero based loops
3. `unchecked`  for limited size collections when incrementing or performing operations with known variables that cannot over/underflow
4. Conversion of some `view` functions to `pure` as no state will be modified (no static calls to precompiles)

### CODE REVIEW

File: backend/plonk/bn254/solidity.go [hunk 1]
```diff
@@ -25,123 +25,141 @@ pragma experimental ABIEncoderV2;
...
-      // sha256(pad || msg || (0 || 48 || 0) || dst || 11)
-      bytes memory tmp;
-      uint8 zero = 0;
-      uint8 lenInBytes = 48;
-      uint8 sizeDomain = 11; // size of dst
-      
-      for (uint i=0; i<64; i++){
-          tmp = abi.encodePacked(tmp, zero);
+  function expand_msg(uint256 x, uint256 y) public pure returns (uint8[48] memory res) {
+    string memory dst = "BSB22-Plonk";
```
Comment (ThomasPiellard) [inline backend/plonk/bn254/solidity.go:38]: Is it possible to make a global constant out of "BSB22-Plonk" also ?

File: backend/plonk/bn254/solidity.go [hunk 2]
```diff
@@ -25,123 +25,141 @@ pragma experimental ABIEncoderV2;
...
+    //uint8[64] memory pad; // 64 is sha256 block size.
+    // sha256(pad || msg || (0 || 48 || 0) || dst || 11)
+    bytes memory tmp;
+    // size of dst
+
+    bytes32 b0;
+    bytes32 b1;
+
+    unchecked {
+      for (uint i; i < 64; ) {
```
Comment (ThomasPiellard) [inline backend/plonk/bn254/solidity.go:49]: So this is on me, but I think the loop could be replaced by
```
uint256 zero_uint256;
tmp = abi.encodePacked(tmp, zero_uint256);
tmp = abi.encodePacked(tmp, zero_uint256);
```

File: backend/plonk/bn254/solidity.go (hunk 1 above)
Comment (thedarkjester) [reply backend/plonk/bn254/solidity.go:38]: indeed it is

File: backend/plonk/bn254/solidity.go (hunk 2 above)
Comment (thedarkjester) [reply backend/plonk/bn254/solidity.go:49]: cool, if it works :) may save even more gas?

**Review by ThomasPiellard — APPROVED** [review]

**Comment by ivokub** [comment]
Closing this pull request. Already included in #781 👍

**Pull request closed without merging**